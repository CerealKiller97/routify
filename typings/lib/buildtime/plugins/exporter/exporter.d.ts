export function exporter({ instance }: {
    instance: any;
}): Promise<[any, any, any, any, any, any, any, any, any, any]>;
export function exportNode(rootNode: RNode, outputDir: string): void;
export function exportInstance(rootNode: RNode, outputDir: string): void;
