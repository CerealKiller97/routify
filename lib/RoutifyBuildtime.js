import { Routify } from '../common/Routify.js'
import { bundlerPlugin } from '../plugins/bundler/index.js'
import { exporterPlugin } from '../plugins/exporter/index.js'
import { filemapperPlugin } from '../plugins/filemapper/index.js'
import { metaFromFilePlugin } from '../plugins/metaFromFile/index.js'

export class RoutifyBuildtime extends Routify {
    plugins = [
        filemapperPlugin,
        metaFromFilePlugin,
        bundlerPlugin,
        exporterPlugin,
    ]
}
