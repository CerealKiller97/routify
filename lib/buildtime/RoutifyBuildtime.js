import '#root/typedef.js'
import { Routify } from '../common/Routify.js'
import { normalizePlugins, sortPlugins, deepAssign } from '../common/utils.js'
import { bundlerPlugin } from './plugins/bundler/index.js'
import { exporterPlugin } from './plugins/exporter/index.js'
import { filemapperPlugin } from './plugins/filemapper/index.js'
import { metaFromFilePlugin } from './plugins/metaFromFile/index.js'
import { watcherPlugin } from './plugins/watcher/index.js'
import { hookHandler, resolvePlugins, sanitizeConfig, throttle } from './utils.js'
import { configent } from 'configent'
import { metaSplitPlugin } from './plugins/metaSplit/index.js'
import fse from 'fs-extra'

const getDefaults = () => ({
    routifyDir: '.routify',
    clearRoutifyDir: true,
    filemapper: {
        moduleFiles: ['_module.svelte', '_reset.svelte'],
        resetFiles: ['_reset.svelte'],
    },
    routesDir: {
        default: 'src/routes',
    },
    extensions: ['.svelte', '.html', '.md', '.svx'],
    plugins: [
        filemapperPlugin,
        metaFromFilePlugin,
        bundlerPlugin,
        metaSplitPlugin,
        exporterPlugin,
        watcherPlugin,
    ],
    watch: false,
})

/**
 * @extends {Routify<RNodeConstructor>}
 */
export class RoutifyBuildtime extends Routify {
    /** @param {RoutifyBuildtimeOptions} options */
    constructor(options) {
        const config = deepAssign(
            {},
            getDefaults(),
            configent({ name: 'routify' }),
            options,
        )
        super(sanitizeConfig(config))

        // normalize routifyDir
        if (typeof this.options.routesDir === 'string')
            this.options.routesDir = { default: options.routesDir }

        // clear dir
        // todo jest doesn't like this one
        if (options.clearRoutifyDir) fse.removeSync(options.routifyDir)

        const _build = this.#build.bind(this)
        this.build = async () => await throttle(_build)
    }

    on = {
        buildStart: hookHandler(),
        buildComplete: hookHandler(),
        fileAdded: hookHandler(),
        fileRemoved: hookHandler(),
        fileChanged: hookHandler(),
        fileWatcherReady: hookHandler(),
    }

    async #build() {
        const instance = this
        this.on.buildStart.runHooks()
        this.nodeIndex.splice(0)

        /** @type {RoutifyBuildtimePlugin[]} */
        const plugins = normalizePlugins(await resolvePlugins(this.options.plugins || []))
        this.plugins = sortPlugins(plugins)
        for (const plugin of this.plugins) {
            const shouldRun =
                typeof plugin.condition === 'undefined' ||
                (await plugin.condition({ instance }))
            if (shouldRun) {
                const { options } = plugin
                await plugin.build({ instance, options })
            }
        }

        this.on.buildComplete.runHooks()
    }

    start = this.#build
}
