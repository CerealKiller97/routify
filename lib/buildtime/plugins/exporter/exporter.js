import { relative, resolve } from 'path'
import {
    createDirname,
    stringifyWithEscape,
    writeFileIfDifferentSync,
} from '../../utils.js'
const __dirname = createDirname(import.meta)

/** @param {{instance: Routify}} param0 */
export const exporter = ({ instance }) => {
    const promises = instance.superNode.children.map(rootNode => {
        rootNode.routifyDir = 'import.meta.url::_EVAL'
        exportNode(rootNode, instance.options.routifyDir)
        exportInstance(rootNode, instance.options.routifyDir)
    })
    return Promise.all(promises)
}

/**
 *
 * @param {RNode} rootNode
 * @param {string} outputDir
 */
export const exportNode = (rootNode, outputDir) => {
    // create imports
    let rootImport = rootNode.meta.bundle
        ? `import * as bundle_${rootNode.id} from './bundles/${rootNode.id}-bundle.js'`
        : ''

    const treeString = stringifyWithEscape(rootNode)

    // prettier-ignore
    const content =
        `${rootImport}` +
        '\n' +
        '\n' +
        `export default ${treeString}`

    const outputPath = resolve(outputDir, `routes.${rootNode.rootName}.js`)
    writeFileIfDifferentSync(outputPath, content)
}

/**
 * creates instance.default.js
 * @param {RNode} rootNode
 * @param {String} outputDir
 */
export const exportInstance = (rootNode, outputDir) => {
    const outputPath = resolve(outputDir, `instance.${rootNode.rootName}.js`)
    const relativeRoutifyPath = relative(
        outputDir,
        resolve(__dirname, '../../../runtime/index.js'),
    ).replace(/\\/g, '/')

    const content = [
        `import { Routify, Router } from '${relativeRoutifyPath}'`,
        `import routes from './routes.${rootNode.rootName}.js'`,
        '',
        'export const instance = new Routify({routes})',
        'export { Router }',
        '',
    ]
    writeFileIfDifferentSync(outputPath, content.join('\n'))
}
