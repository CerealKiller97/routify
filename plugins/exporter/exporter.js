import { Node } from '../../common/Node.js' //eslint-disable-line
import { Routify } from '../../common/Routify.js' //eslint-disable-line
import { relative, resolve } from 'path'
import fse from 'fs-extra'

/** @param {{instance: Routify}} param0 */
export const exporter = ({ instance }) => {
    const promises = instance.superNode.children.map((rootNode) =>
        exportNode(rootNode, instance.options.routifyDir),
    )
    return Promise.all(promises)
}

/**
 * JSON.stringify wrapper that leaves specified keys unquoted
 * @param {any} obj
 * @param {string[]} keys
 * @returns {string}
 */
const stringifyWithEscape = (obj, keys) => {
    const re = new RegExp(`^( *)"(${keys.join('|')})": (".+")`, 'gm')
    return JSON.stringify(obj, null, 2).replace(
        re,
        (m, spaces, key, value) => `${spaces}"${key}": ${JSON.parse(value)}`,
    )
}

/**
 *
 * @param {Node} rootNode
 * @param {string} outputDir
 */
export const exportNode = (rootNode, outputDir) => {
    // create imports
    const imports = [rootNode, ...rootNode.descendants]
        .filter(
            (node) => node.component && !node.component.startsWith('import('),
        )
        .map(
            (node) =>
                `import ${node.id} from '${relative(
                    outputDir,
                    node.component,
                ).replace(/\\/g, '/')}'`,
        )
        .join('\n')

    // set component to id
    ;[rootNode, ...rootNode.descendants]
        .filter(
            (node) => node.component && !node.component.startsWith('import('),
        )
        .forEach((node) => (node.component = node.id))

    const treeString = stringifyWithEscape(rootNode.map, ['component'])
    const outputPath = resolve(outputDir, `routes.${rootNode.rootName}.js`)
    const content = [imports, `export const routes = ${treeString}`].join(
        '\n\n',
    )

    fse.outputFileSync(outputPath, content)
}
