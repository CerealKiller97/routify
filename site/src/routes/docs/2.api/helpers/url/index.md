<script>
    import { resolveNode } from '@roxi/routify'
    import FilesViewer from '#cmp/FilesViewer.svelte'
    import Example from '#cmp/Example.svelte'
</script>


The `$url` helper creates URLs from paths and parameters.

<FilesViewer root={resolveNode('../example')} focus="index.svelte" />

---

# URL Example
<Example offset="../example" />
