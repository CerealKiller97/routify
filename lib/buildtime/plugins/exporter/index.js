import '../../../../typedef.js'
import { exporter } from './exporter.js'

/** @type {RoutifyBuildtimePlugin} */
export const exporterPlugin = {
    name: 'exporter',
    after: 'filemapper',
    build: exporter,
}
