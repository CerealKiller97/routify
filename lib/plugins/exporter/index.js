import '#root/typedef.js'
import { exporter } from './exporter.js'

/** @type {RoutifyPlugin} */
export const exporterPlugin = {
    condition: () => true,
    after: 'filemapper',
    mode: 'compile',
    run: exporter,
}
