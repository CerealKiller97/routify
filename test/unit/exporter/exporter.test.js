import { suite } from 'uvu'
import fse from 'fs-extra'
import * as assert from 'uvu/assert'
import { dirname, resolve } from 'path'
import { fileURLToPath } from 'url'
import { exportNode } from '../../../plugins/exporter/exporter.js'
import { filemapper } from '../../../plugins/filemapper/lib/index.js'
import { RoutifyBuildtime } from '../../../lib/RoutifyBuildtime.js'

const __dirname = dirname(fileURLToPath(import.meta.url))
const test = suite('exporter')
const options = {
    filemapper: {
        routesDir: { default: __dirname + '/example' },
    },
}

const instance = new RoutifyBuildtime(options)

test('can build route tree', async () => {
    await filemapper({ instance })
    assert.is(instance.superNode.children[0].descendants.length, 7)
})

test('can export a route tree', async () => {
    await exportNode(
        instance.superNode.children[0],
        __dirname + '/temp/.routify',
    )
    const content = fse.readFileSync(
        resolve(__dirname, 'temp/.routify', 'routes.default.js'),
        'utf-8',
    )
    assertSnapshot('routes.js', content, 0)
})

test.run()

function assertSnapshot(name, content, update) {
    content = JSON.parse(JSON.stringify(content))
    const filepath = `${__dirname}/fixtures/${name}`
    if (update) fse.outputFileSync(filepath, content)
    const expect = fse.readFileSync(filepath, 'utf-8')
    assert.snapshot(content, expect)
}
