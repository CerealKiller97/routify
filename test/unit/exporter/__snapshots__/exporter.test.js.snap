// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can export a route tree 1`] = `
"import * as _default from '../../example/_module.svelte'
import * as _default_admin from '../../example/admin/_reset.svelte'
import * as _default_index_svelte from '../../example/index.svelte'
import * as _default_admin_index_svelte from '../../example/admin/index.svelte'
import * as _default_admin_page_svelte from '../../example/admin/page.svelte'
import * as _default_blog__slug__svelte from '../../example/blog/[slug].svelte'
import * as _default_blog_prefixed_slug__svelte from '../../example/blog/prefixed[slug].svelte'

export const routes = {
  \\"id\\": \\"_default\\",
  \\"component\\": () => _default,
  \\"meta\\": {},
  \\"rootName\\": \\"default\\",
  \\"file\\": {
    \\"path\\": \\"test/unit/exporter/example/_module.svelte\\",
    \\"dir\\": \\"test/unit/exporter/example\\",
    \\"base\\": \\"_module.svelte\\",
    \\"ext\\": \\".svelte\\",
    \\"name\\": \\"_module\\"
  },
  \\"children\\": [
    {
      \\"id\\": \\"_default_admin\\",
      \\"name\\": \\"admin\\",
      \\"component\\": () => _default_admin,
      \\"meta\\": {
        \\"reset\\": {
          \\"value\\": true
        }
      },
      \\"file\\": {
        \\"path\\": \\"test/unit/exporter/example/admin/_reset.svelte\\",
        \\"dir\\": \\"test/unit/exporter/example/admin\\",
        \\"base\\": \\"_reset.svelte\\",
        \\"ext\\": \\".svelte\\",
        \\"name\\": \\"_reset\\"
      },
      \\"children\\": [
        {
          \\"id\\": \\"_default_admin_index_svelte\\",
          \\"name\\": \\"index\\",
          \\"component\\": () => _default_admin_index_svelte,
          \\"meta\\": {},
          \\"file\\": {
            \\"path\\": \\"test/unit/exporter/example/admin/index.svelte\\",
            \\"dir\\": \\"test/unit/exporter/example/admin\\",
            \\"base\\": \\"index.svelte\\",
            \\"ext\\": \\".svelte\\",
            \\"name\\": \\"index\\"
          },
          \\"children\\": []
        },
        {
          \\"id\\": \\"_default_admin_page_svelte\\",
          \\"name\\": \\"page\\",
          \\"component\\": () => _default_admin_page_svelte,
          \\"meta\\": {},
          \\"file\\": {
            \\"path\\": \\"test/unit/exporter/example/admin/page.svelte\\",
            \\"dir\\": \\"test/unit/exporter/example/admin\\",
            \\"base\\": \\"page.svelte\\",
            \\"ext\\": \\".svelte\\",
            \\"name\\": \\"page\\"
          },
          \\"children\\": []
        }
      ]
    },
    {
      \\"id\\": \\"_default_blog\\",
      \\"name\\": \\"blog\\",
      \\"component\\": false,
      \\"meta\\": {},
      \\"file\\": {
        \\"path\\": \\"test/unit/exporter/example/blog\\",
        \\"dir\\": \\"test/unit/exporter/example\\",
        \\"base\\": \\"blog\\",
        \\"ext\\": \\"\\",
        \\"name\\": \\"blog\\"
      },
      \\"children\\": [
        {
          \\"id\\": \\"_default_blog__slug__svelte\\",
          \\"name\\": \\"[slug]\\",
          \\"component\\": () => _default_blog__slug__svelte,
          \\"meta\\": {
            \\"dynamic\\": {
              \\"value\\": true
            }
          },
          \\"file\\": {
            \\"path\\": \\"test/unit/exporter/example/blog/[slug].svelte\\",
            \\"dir\\": \\"test/unit/exporter/example/blog\\",
            \\"base\\": \\"[slug].svelte\\",
            \\"ext\\": \\".svelte\\",
            \\"name\\": \\"[slug]\\"
          },
          \\"children\\": []
        },
        {
          \\"id\\": \\"_default_blog_prefixed_slug__svelte\\",
          \\"name\\": \\"prefixed[slug]\\",
          \\"component\\": () => _default_blog_prefixed_slug__svelte,
          \\"meta\\": {
            \\"dynamic\\": {
              \\"value\\": true
            }
          },
          \\"file\\": {
            \\"path\\": \\"test/unit/exporter/example/blog/prefixed[slug].svelte\\",
            \\"dir\\": \\"test/unit/exporter/example/blog\\",
            \\"base\\": \\"prefixed[slug].svelte\\",
            \\"ext\\": \\".svelte\\",
            \\"name\\": \\"prefixed[slug]\\"
          },
          \\"children\\": []
        }
      ]
    },
    {
      \\"id\\": \\"_default_index_svelte\\",
      \\"name\\": \\"index\\",
      \\"component\\": () => _default_index_svelte,
      \\"meta\\": {},
      \\"file\\": {
        \\"path\\": \\"test/unit/exporter/example/index.svelte\\",
        \\"dir\\": \\"test/unit/exporter/example\\",
        \\"base\\": \\"index.svelte\\",
        \\"ext\\": \\".svelte\\",
        \\"name\\": \\"index\\"
      },
      \\"children\\": []
    }
  ]
}"
`;
